### PR TITLE
Add generic dependencies

### PR DESCRIPTION
**Pull request by mheon**
Right now, libpod has a notion of dependencies between containers - if container A uses container B's network namespace, container B must be running before container A can start so we can join that namespace. These patches extend this notion of dependencies to be generic - we can now make a container depend on other containers, only starting when these other containers are started. This allows enforced ordering of container start.

I'm working on patches to add a `--recursive` flag to `start` and `remove` to take advantage of this, starting a container and all dependencies that are not running, and removing a container and all its dependencies. Those will come in a separate PR so as to not clutter things.

### CODE REVIEW

File: cmd/podman/common.go [hunk 1]
```diff
@@ -137,6 +137,10 @@ var createFlags = []cli.Flag{
 		Name:  "cpuset-mems",
 		Usage: "Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.",
 	},
+	cli.StringSliceFlag{
+		Name:  "depends-on",
+		Usage: "Add dependency containers which must be running before this container starts",
```
Comment (rhatdan) [inline cmd/podman/common.go]: How is a human going to know this?  Not sure what this even means?  I am not sure we want to expose this or support it, maybe should be hidden?

Comment (mheon) [reply cmd/podman/common.go]: I was mostly thinking of ordering service startup - ensuring that the DB container is started before the HTTP server container spins up. This isn't good microservices practice, but I can see this as an easy way establishing service ordering without getting into systemd unit files.

**Comment by mheon** [comment]
bot, retest this please

File: cmd/podman/common.go (hunk 1 above)
Comment (rhatdan) [reply cmd/podman/common.go]: Those are jobs for systemd not for podman.
I could see this patch set becoming very complex and ordering is notoriously bad.  You would have to worry about circular dependencies.  ETC. I don't think this is a good idea.  Leave it to Systemd.

Comment (mheon) [reply cmd/podman/common.go]: We already have to worry about those things for namespace dependencies - that's why we have the dependency graph generation code we use for pod startup.

Comment (rhatdan) [reply cmd/podman/common.go]: Sure, but that is the definition of the pod, but I don't see this as adding anything but headaches, and would be something we would want to discourage.  As soon as you get into complex dependencies, like this there are lots of errors.  

I don't see us advising users to use this, we would advise them to setup services to run in unit files with proper dependencies defined in systemd. Then they would always be defining them in the same place.

Comment (mheon) [reply cmd/podman/common.go]: If you want, we can restrict it to only be used within pods at the CLI level, but the underlying code is generic enough that we can use it without pods if we want to. Within pods, the typical example would be to depend on the pause container in Kube pods, but I'm sure people can think of other ways of using this.

Comment (rhatdan) [reply cmd/podman/common.go]: Sure I can think of ways people would use it, but this is not something we want to support.

Comment (mheon) [reply cmd/podman/common.go]: If you want, I can scrap the CLI frontend. The backend code is sufficiently trivial/useful that I'd like to leave it in. I also think start/remove --recursive are useful enough to justify (`podman remove --recursive --force` is needed to guarantee we kick out a container right now)

Comment (rhatdan) [reply cmd/podman/common.go]: Remove the CLI, and then I am less concerned.  If we find the need we could eventually add it back.
Not sure you need --recursive.  Perhaps --force should be enough.  I am not sure what docker does in this situation.  Does it fail to stop a container if a second container relies on it, or does it kill the first and the second keeps open the namespaces that are shared.   

podman rm --force
CID=podman run -d ...
podman run -d --net  container:$CID --PID container $CID ...
podman stop $CID
What happens?

podman rm $CID --force

What happens?

**Comment by mheon** [comment]
Docker allows you to remove the container you depend on. If you start the container with the dependency again, it will fail, saying the container it depends on no longer exists.

We treat dependencies a little more seriously (if the CRI-O pause container gets deleted by someone fiddling with `podman rm` that's not a good thing), so we won't let you remove the container if things depend on it, but this does cause issues (`rm -f` no longer guarantees a container is gone). I think a `recursive` flag is a good solution here, because it makes it clear other containers may be gone after you remove this one.

**Comment by rhatdan** [comment]
I believe -f should imply --recursive.  If I say I want a container or Pod gone and I say force it, then all of the containers that rely on the container/pod should be removed also.

**Comment by mheon** [comment]
@rhatdan Dropped the commit with the CLI parts. There are still a few changes to `podman run` and `podman start`, but it's mostly simplifying things by removing `Init()` before `Start()`.

On --force implying recursive: I don't know if I agree, but we can have that discussion once the code is actually written...

**Comment by mheon** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
LGTM

**Comment by rhatdan** [comment]
@baude @umohnani8 @TomSweeneyRedHat PTAL

**Comment by umohnani8** [comment]
LGTM

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**